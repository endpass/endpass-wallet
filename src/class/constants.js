export const REQUEST_TIMEOUT_MSEC = 30000;

export default {
  REQUEST_TIMEOUT_MSEC,
};
