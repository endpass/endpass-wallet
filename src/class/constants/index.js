export { INPAGE_EVENT, INPAGE_ID_PREFIX } from './InpageProvider';

export const REQUEST_TIMEOUT_MSEC = 30000;

export const STORAGE_USER_META_KEY = 'endpass-user-meta';

export const AVAILABLE_USER_META_PROPS = ['activeAccount'];

export const PROXY_REQUEST_PREFIX = 'endpass-identity';

export const HARDWARE_DERIVIATION_PATH = `m/44'/60'/0'/0/`;

export default {
  REQUEST_TIMEOUT_MSEC,
  STORAGE_USER_META_KEY,
  AVAILABLE_USER_META_PROPS,
  PROXY_REQUEST_PREFIX,
  HARDWARE_DERIVIATION_PATH,
};
