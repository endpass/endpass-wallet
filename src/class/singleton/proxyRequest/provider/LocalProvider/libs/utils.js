import regPath from 'path-to-regexp';

const utils = {
  getRouteKey(routesRegexp, routePath) {
    return Object.keys(routesRegexp).find(key => {
      const re = routesRegexp[key];
      return re.test(routePath);
    });
  },
  getPath(serverUrl, url) {
    const levels = url.split(serverUrl);
    const keysPath = levels.pop() || '';
    const ret = keysPath[0] === '/' ? keysPath.substring(1) : keysPath;

    return ret;
  },
  getPathArgs(routesRegexp, key, routePath) {
    const keys = regPath.parse(key);
    const re = routesRegexp[key];
    re.lastIndex = 0; // drop exec position

    const values = re.exec(routePath);
    return keys.reduce((map, item, index) => {
      if (item && item.name) {
        map[item.name] = values[index];
      }
      return map;
    }, {});
  },
};

export default utils;
