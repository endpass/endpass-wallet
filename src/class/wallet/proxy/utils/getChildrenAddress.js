export default function(hdWallet, offset, limit) {
  return [...Array(limit)].map((_, i) =>
    hdWallet
      .deriveChild(offset + i)
      .getWallet()
      .getChecksumAddressString(),
  );
}
