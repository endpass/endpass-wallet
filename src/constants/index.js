export { INPAGE_EVENT, INPAGE_ID_PREFIX } from './InpageProvider';

export const MAIN_NET_ID = 1;

export const NETWORK_URL = Object.freeze({
  ETH: [
    'wss://eth-mainnet.endpass.com:2084',
    'wss://eth-mainnet.endpass.com',
    'https://eth-mainnet.endpass.com:2083',
    `https://mainnet.infura.io/${ENV.infuraConf.key}`,
  ],
  ROP: [
    'wss://eth-ropsten.endpass.com:2084',
    'wss://eth-ropsten.endpass.com',
    'https://eth-ropsten.endpass.com:2083',
    `https://ropsten.infura.io/${ENV.infuraConf.key}`,
  ],
  RIN: `https://rinkeby.infura.io/${ENV.infuraConf.key}`,
  ETC: [
    'wss://etc-mainnet.endpass.com:2084',
    'wss://etc-mainnet.endpass.com',
    'https://etc-mainnet.endpass.com:2083',
    'https://etc-geth.0xinfra.com',
  ],
});

export const DEFAULT_NETWORKS = Object.freeze([
  {
    id: 1,
    networkType: 'main',
    currency: 1,
    name: 'Main',
    url: NETWORK_URL.ETH,
  },
  {
    id: 3,
    name: 'Ropsten',
    networkType: 'ropsten',
    currency: 2,
    url: NETWORK_URL.ROP,
  },
  {
    id: 4,
    name: 'Rinkeby',
    networkType: 'rinkeby',
    currency: 2,
    url: NETWORK_URL.RIN,
  },
  {
    id: 61,
    name: 'Ethereum classic',
    networkType: 'ethClassic',
    currency: 3,
    url: NETWORK_URL.ETC,
  },
]);

export const CURRENCIES = Object.freeze([
  {
    name: 'ETH',
    id: 1,
  },
  {
    name: 'ETH-TEST',
    id: 2,
  },
  {
    name: 'ETC',
    id: 3,
  },
]);

export const AVAILABLE_FIAT_CURRENCIES = Object.freeze(
  // prettier-ignore
  ['USD', 'AUD', 'BRL', 'CAD', 'CHF', 'CLP', 'CNY', 'CZK', 'DKK', 'EUR', 'GBP', 'HKD', 'HUF', 'IDR', 'ILS', 'INR', 'JPY', 'KRW', 'MXN', 'MYR', 'NOK', 'NZD', 'PHP', 'PKR', 'PLN', 'RUB', 'SEK', 'SGD', 'THB', 'TRY', 'TWD', 'ZAR'],
);

export const IDENTITY_MODE = Object.freeze({
  DEFAULT: 'default',
  CUSTOM: 'custom',
  LOCAL: 'local',
});

export const PROXY_REQUEST_PREFIX = 'endpass-identity';

export const STORAGE_USER_META_KEY = 'endpass-user-meta';

export const AVAILABLE_USER_META_PROPS = ['activeAccount'];

export const REQUEST_TIMEOUT_MSEC = 30000;

export const HARDWARE_DERIVIATION_PATH = `m/44'/60'/0'/0/`;

export const HARDWARE_WALLET_TYPE = Object.freeze({
  TREZOR: 'TrezorAccount',
  LEDGER: 'LedgerAccount',
});

export const WALLET_TYPE = Object.freeze({
  PUBLIC: 'PublicAccount',
  HD_PUBLIC: 'HDPublicAccount',
  HD_MAIN: 'HDMainAccount',
  ...HARDWARE_WALLET_TYPE,
});

export const DAPP_WHITELISTED_METHODS = [
  'personal_sign',
  'eth_personalSign',
  'eth_signTypedData',
  'eth_sendTransaction',
];

export default {
  AVAILABLE_FIAT_CURRENCIES,
  MAIN_NET_ID,
  DEFAULT_NETWORKS,
  CURRENCIES,
  IDENTITY_MODE,
  PROXY_REQUEST_PREFIX,
  STORAGE_USER_META_KEY,
  AVAILABLE_USER_META_PROPS,
  REQUEST_TIMEOUT_MSEC,
  HARDWARE_DERIVIATION_PATH,
  HARDWARE_WALLET_TYPE,
  WALLET_TYPE,
  DAPP_WHITELISTED_METHODS,
};
