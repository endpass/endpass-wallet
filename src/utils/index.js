import store from '../store';

export { default as http } from './http';

export const kebabToCamel = input =>
  input.replace(/-([a-z])/g, g => g[1].toUpperCase());

export const getInitializedValueFromStore = (module, field) => {
  return new Promise(resolve => {
    const value = module[field];

    if (value === null) {
      const unwatch = store.watch(
        () => module[field],
        newValue => {
          unwatch();
          resolve(newValue);
        },
      );
    } else {
      resolve(value);
    }
  });
};
