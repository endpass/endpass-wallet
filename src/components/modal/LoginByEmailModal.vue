<template>
  <v-modal
    class="is-dark"
    @close="handleClose"
  >
    <template slot="header">Get Started</template>

    <p class="subtitle">Please enter your email address below to access your
    wallet or create a new one.</p>

    <v-form
      id="loginByEmail"
      v-model="isFormValid"
      @submit="handleSubmit"
    >
      <v-input
        v-model="email"
        :disabled="isLoading"
        label="Email"
        help="Your email address may be used to help recover your
               wallet in case you lose access."
        name="email"
        validator="required|email"
        placeholder="Your email"
      />

      <v-select
        :disabled="isLoading"
        :options="availableIdentityServerTypes"
        v-model="currentIdentityServerType"
        label="Identity Server"
        name="currentIdentityServerType"
      />

      <v-input
        v-if="selectedCustomIdentityServerType"
        id="customIdentityServer"
        v-model="customIdentityServer"
        :disabled="isLoading"
        label="Custom Identity Server"
        name="customIdentityServer"
        validator="required|url:require_protocol:true"
        placeholder="Custom Identity Server"
      />

      <v-checkbox v-model="termsAccepted">
        I accept the <a
          href="https://endpass.com/terms/"
          target="_blank"
        >
          Terms of Service
        </a>
        and
        <a
          href="https://endpass.com/privacy/"
          target="_blank"
        >
          Privacy Policy
        </a>.
      </v-checkbox>
    </v-form>
    <div
      slot="footer"
      class="buttons"
    >
      <v-button
        :disabled="!termsAccepted || !isFormValid"
        :loading="isLoading"
        class-name="is-primary is-medium"
        form="loginByEmail"
        data-test="submit-login"
      >
        Continue
      </v-button>
    </div>
  </v-modal>
</template>

<script>
import VModal from '@/components/ui/VModal';
import VForm from '@/components/ui/form/VForm';
import VInput from '@/components/ui/form/VInput';
import VButton from '@/components/ui/form/VButton';
import VCheckbox from '@/components/ui/form/VCheckbox';
import VSelect from '@/components/ui/form/VSelect';
import { IDENTITY_MODE } from '@/constants';

const availableIdentityServerTypes = [
  { text: 'Endpass', val: IDENTITY_MODE.DEFAULT },
  { text: 'Local Storage', val: IDENTITY_MODE.LOCAL },
  { text: 'Custom server', val: IDENTITY_MODE.CUSTOM },
];

export default {
  name: 'LoginByEmailModal',
  props: {
    isLoading: {
      type: Boolean,
      default: false,
    },
  },
  data: () => ({
    email: '',
    termsAccepted: true,
    availableIdentityServerTypes,
    currentIdentityServerType: availableIdentityServerTypes[0].val,
    customIdentityServer: null,
    isFormValid: false,
  }),
  computed: {
    selectedCustomIdentityServerType() {
      return this.currentIdentityServerType === IDENTITY_MODE.CUSTOM;
    },
  },
  methods: {
    handleSubmit() {
      this.$ga.event({
        eventCategory: 'onboarding',
        eventAction: 'submit_email',
      });

      const { email } = this;
      const mode = {
        type: this.currentIdentityServerType,
        serverUrl: this.selectedCustomIdentityServerType
          ? this.customIdentityServer
          : undefined,
      };

      this.$emit('confirm', { email, mode });
    },
    handleClose() {
      this.$emit('close');
    },
  },
  components: {
    VModal,
    VForm,
    VInput,
    VButton,
    VCheckbox,
    VSelect,
  },
};
</script>
