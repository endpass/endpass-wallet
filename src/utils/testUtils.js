import { kebabCase } from 'lodash';
import { getFrom } from '@/utils/objects';

function convertDirectivesToAttrs(directives) {
  const attrs = {};

  if (directives) {
    directives.forEach(({ rawName, expression }) => {
      attrs[rawName] = expression;
    });
  }

  return attrs;
}

function convertListenersToAttrs(listeners) {
  const attrs = {};

  if (listeners) {
    Object.keys(listeners).forEach(key => {
      const handlers = [].concat(listeners[key]);
      const handler = handlers.find(h => {
        const name = getFrom(h, 'fns.name', 'name') || '';

        return name.includes('bound ');
      });

      if (handler) {
        attrs[`v-on:${key}`] = (handler.fns || handler).name;
      }
    });
  }

  return attrs;
}

function convertModelToAttrs(model) {
  return model ? { 'v-model': model.expression } : {};
}

function convertSlotToAttrs(slot) {
  return slot ? { slot } : {};
}

function generateElement(createElement, vnode) {
  return vnode.componentOptions
    ? generateElementFromComponent(createElement, vnode)
    : generateElementFromHTML(createElement, vnode);
}

function getChildrenElements(createElement, children) {
  return children
    ? children.map(child => generateElement(createElement, child))
    : undefined;
}

function generateElementFromComponent(
  createElement,
  { data, componentOptions },
) {
  const { attrs, directives, model } = data;
  const { tag, listeners, children } = componentOptions;
  const componentData = {
    ...data,
    attrs: {
      ...attrs,
      ...convertDirectivesToAttrs(directives),
      ...convertListenersToAttrs(listeners),
      ...convertModelToAttrs(model),
      class: data.staticClass,
    },
    hook: undefined,
    staticClass: undefined,
  };

  if (listeners) {
    componentData.on = Object.assign({}, listeners);
  }

  return createElement(
    tag,
    componentData,
    getChildrenElements(createElement, children),
  );
}

function generateElementFromHTML(createElement, vnode) {
  if (!vnode.data) {
    return vnode;
  }

  const { tag, data, children } = vnode;
  const { directives, on, model, attrs = {}, slot } = data;

  return createElement(
    tag,
    {
      ...data,
      attrs: {
        ...attrs,
        ...convertDirectivesToAttrs(directives),
        ...convertListenersToAttrs(on),
        ...convertModelToAttrs(model),
        ...convertSlotToAttrs(slot),
      },
    },
    getChildrenElements(createElement, children),
  );
}

export function generateStubs(Component) {
  return Object.values(Component.components).reduce((stubs, stubComponent) => {
    let elementName;

    if (stubComponent.name) {
      elementName = kebabCase(stubComponent.name);
    }

    return Object.assign(stubs, {
      [elementName]: {
        render(createElement) {
          return generateElementFromComponent(createElement, this.$vnode);
        },
      },
    });
  }, {});
}

export default {
  generateStubs,
};
