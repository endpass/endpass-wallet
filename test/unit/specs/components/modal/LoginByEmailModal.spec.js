import Vuex from 'vuex';
import VeeValidate from 'vee-validate';
import { mount, shallow, createLocalVue } from '@vue/test-utils';

import { IDENTITY_MODE } from '@/constants';
import LoginByEmailModal from '@/components/modal/LoginByEmailModal';
import { generateStubs } from '@/utils/testUtils';

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VeeValidate);

describe('LoginByEmailModal', () => {
  let wrapper;

  beforeEach(() => {
    const $ga = { event: jest.fn() };
    const store = new Vuex.Store({
      modules: {
        user: {
          namespaced: true,
          actions: {
            validateCustomServer: jest.fn().mockResolvedValue(true),
          },
        },
      },
    });

    wrapper = shallow(LoginByEmailModal, {
      store,
      localVue,
      stubs: generateStubs(LoginByEmailModal),
      mocks: {
        $ga,
      },
    });
  });

  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('LoginByEmailModal');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });

    it('should display a field for custom identity server', () => {
      wrapper.setData({
        currentIdentityServerType: IDENTITY_MODE.CUSTOM,
      });

      expect(wrapper.find('#customIdentityServer').html()).toMatchSnapshot();
    });
  });

  describe('behavior', () => {
    describe('allow input', () => {
      it('should allow input by default', () => {
        expect(wrapper.vm.isInputAllowed).toBe(true);
      });

      it('should not allow input when loading', () => {
        wrapper.setProps({ isLoading: true });

        expect(wrapper.vm.isInputAllowed).toBe(false);
      });

      it('should not allow input when validating', () => {
        wrapper.setData({ isValidating: true });

        expect(wrapper.vm.isInputAllowed).toBe(false);
      });
    });

    describe('handle submit modal', () => {
      const email = 'email';
      const customIdentityServer = 'http://custom.com/api///';
      const validCustomIdentityServer = customIdentityServer.replace(
        /\/+$/,
        '',
      );

      it('should trigger "confirm" event', async () => {
        expect.assertions(1);

        const currentIdentityServerType = IDENTITY_MODE.DEFAULT;
        const expected = [
          {
            email,
            mode: {
              type: currentIdentityServerType,
              serverUrl: undefined,
            },
          },
        ];

        wrapper.setData({ email, currentIdentityServerType });

        await wrapper.vm.handleSubmit();

        expect(wrapper.emitted().confirm).toEqual([expected]);
      });

      it('should trigger "confirm" event with custom identity server url', async () => {
        expect.assertions(1);

        const currentIdentityServerType = IDENTITY_MODE.CUSTOM;
        const expected = [
          {
            email,
            mode: {
              type: currentIdentityServerType,
              serverUrl: validCustomIdentityServer,
            },
          },
        ];

        wrapper.setData({
          email,
          currentIdentityServerType,
          customIdentityServer,
        });

        await wrapper.vm.handleSubmit();

        expect(wrapper.emitted().confirm).toEqual([expected]);
      });

      it('should validate the custom server if the identity mode is custom', async () => {
        expect.assertions(2);

        wrapper.vm.validateServer = jest.fn();

        wrapper.setData({
          currentIdentityServerType: IDENTITY_MODE.CUSTOM,
          customIdentityServer,
        });

        await wrapper.vm.handleSubmit();

        expect(wrapper.vm.validateServer).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.validateServer).toBeCalledWith(
          validCustomIdentityServer,
        );
      });

      it('should not validate the custom server if the identity mode is not custom', async () => {
        expect.assertions(1);

        wrapper.setData({
          currentIdentityServerType: IDENTITY_MODE.DEFAULT,
        });
        wrapper.vm.validateServer = jest.fn();

        await wrapper.vm.handleSubmit();

        expect(wrapper.vm.validateServer).not.toBeCalled();
      });

      it('should emit error if the custom identity server if invalid', async () => {
        expect.assertions(2);

        const error = new Error();

        wrapper.setData({
          currentIdentityServerType: IDENTITY_MODE.CUSTOM,
          customIdentityServer,
        });
        wrapper.vm.validateServer = jest.fn().mockRejectedValueOnce(error);
        wrapper.vm.emitError = jest.fn();

        await wrapper.vm.handleSubmit();

        expect(wrapper.vm.emitError).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.emitError).toBeCalledWith(error);
      });
    });

    describe('handle close modal', () => {
      it('should trigger "close" event', () => {
        wrapper.vm.handleClose();

        expect(wrapper.emitted().close).toEqual([[]]);
      });
    });

    describe('custom server validation', () => {
      const serverUrl = 'server';

      it('should validate the custom server', async () => {
        expect.assertions(2);

        wrapper.vm.validateCustomServer = jest.fn();

        await wrapper.vm.validateServer(serverUrl);

        expect(wrapper.vm.validateCustomServer).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.validateCustomServer).toBeCalledWith({ serverUrl });
      });

      it('should activate the loader while validating a custom server', async () => {
        expect.assertions(2);

        jest.useFakeTimers();

        wrapper.vm.validateCustomServer = jest.fn(
          () =>
            new Promise(res => {
              setTimeout(() => {
                res();
              }, 1000);
            }),
        );

        wrapper.vm.validateServer();
        await wrapper.vm.$nextTick();

        expect(wrapper.vm.isValidating).toBe(true);

        jest.runAllTimers();
        await wrapper.vm.$nextTick();

        expect(wrapper.vm.isValidating).toBe(false);
      });
    });

    describe('validation', () => {
      beforeEach(() => {
        wrapper = mount(LoginByEmailModal, {
          localVue,
        });
      });

      it('should disable submit button by default', () => {
        const button = wrapper.find('[data-test=submit-login]');
        expect(button.attributes().disabled).toBeTruthy();
      });
    });
  });
});
