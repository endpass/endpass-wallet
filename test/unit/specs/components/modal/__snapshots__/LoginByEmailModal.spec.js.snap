// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginByEmailModal render should display a field for custom identity server 1`] = `<v-input id="customIdentityServer" label="Custom Identity Server" name="customIdentityServer" validator="required|url:require_protocol:true" placeholder="Custom Identity Server" help="Example: https://yourserver.com/api" v-model="customIdentityServer"></v-input>`;

exports[`LoginByEmailModal render should render component 1`] = `
<v-modal
  class="is-dark"
  v-on:close="bound handleClose"
>
  <template
    slot="header"
  >
    Get Started
  </template>
   
  <p
    class="subtitle"
  >
    
    Please enter your email address below to access your wallet or create a new one.
  
  </p>
   
  <v-form
    id="loginByEmail"
    v-model="isFormValid"
    v-on:submit="bound handleSubmit"
  >
    <v-input
      help="Your email address may be used to help recover your
             wallet in case you lose access."
      label="Email"
      name="email"
      placeholder="Your email"
      v-model="email"
      validator="required|email"
    />
     
    <v-select
      label="Identity Server"
      name="currentIdentityServerType"
      options="[object Object],[object Object],[object Object]"
      v-model="currentIdentityServerType"
    />
     
    <!---->
     
    <v-checkbox
      v-model="termsAccepted"
    >
      
      I accept the 
      <a
        href="https://endpass.com/terms/"
        target="_blank"
      >
        
        Terms of Service
      
      </a>
      
      and
      
      <a
        href="https://endpass.com/privacy/"
        target="_blank"
      >
        
        Privacy Policy
      
      </a>
      .
    
    </v-checkbox>
  </v-form>
   
  <div
    class="buttons"
    slot="footer"
  >
    <v-button
      class-name="is-primary is-medium"
      data-test="submit-login"
      disabled="true"
      form="loginByEmail"
    >
      
      Continue
    
    </v-button>
  </div>
</v-modal>
`;
